base/rational_to_float.jl

```diff
@@ -0,0 +1,214 @@
+    excess_width = width - prec_p_1
+    exp = width - shift - true
+
+    exp_underflow = if isnothing(max_subnormal_exp)
+        zero(exp)
+    else
+        let d = max_subnormal_exp - exp, T = typeof(d), z = zero(d)::T
+            (signbit(d) ? z : d + true)::T
+        end
+    end
+
+    (quo1, rem1) = divrem_2(quo0, exp_underflow + excess_width)
+    integral_significand = quo1 >>> true
+    round_bit = quo1 % Bool
+    sticky_bit = !iszero(rem1) | !iszero(rem0)
+
+    (; integral_significand, exponent = exp, round_bit, sticky_bit)
+end
+
+struct RoundingIncrementHelper
+    final_bit::Bool
+    round_bit::Bool
+    sticky_bit::Bool
+end
+
+(h::RoundingIncrementHelper)(::Rnd.FinalBit) = h.final_bit
+(h::RoundingIncrementHelper)(::Rnd.RoundBit) = h.round_bit
+(h::RoundingIncrementHelper)(::Rnd.StickyBit) = h.sticky_bit
+
+function to_float_components_rounded(num, den, precision, max_subnormal_exp, romo, sign_bit)
+    overflows = (x, p) -> x == machine_shift(<<, one(x), p)
+    t = to_float_components_impl(num, den, precision, max_subnormal_exp)
+    raw_significand = t.integral_significand
+    rh = RoundingIncrementHelper(raw_significand % Bool, t.round_bit, t.sticky_bit)
+    incr = Rnd.correct_rounding_requires_increment(rh, romo, sign_bit)
+    rounded = raw_significand + incr
+    (integral_significand, exponent) = let exp = t.exponent
+        if overflows(rounded, precision)
+            (rounded >>> true, exp + true)
+        else
+            (rounded, exp)
+        end
+    end
+    (; integral_significand, exponent)
+end
+
+function to_float_components(::Type{T}, num, den, precision, max_subnormal_exp, romo, sb) where {T}
+    to_float_components_rounded(abs(T(num)), den, precision, max_subnormal_exp, romo, sb)
+end
+
+function to_floating_point_fallback(::Type{T}, ::Type{S}, num, den, rm, prec) where {T,S}
+    num_is_zero = iszero(num)
+    den_is_zero = iszero(den)
+    sb = signbit(num)
+    is_zero = num_is_zero & !den_is_zero
+    is_inf = !num_is_zero & den_is_zero
+    is_regular = !num_is_zero & !den_is_zero
+    if is_regular
+        let
+            c = to_float_components(S, num, den, prec, nothing, rm, sb)
+            exp = c.exponent
+            signif = T(c.integral_significand)::T
+            let x = ldexp(signif, exp - prec + true)::T
+                sb ? -x : x
+            end::T
+        end
+    else
+        if is_zero
+            zero(T)::T
+        elseif is_inf
+            T(Inf)::T
+        else
+            T(NaN)::T
+        end
+    end::T
+end
+
+function to_floating_point_impl(::Type{T}, ::Type{S}, num, den, rm, prec) where {T,S}
+    to_floating_point_fallback(T, S, num, den, rm, prec)
+end
+
+function to_floating_point_impl(::Type{T}, ::Type{S}, num, den, rm, prec) where {T<:Base.IEEEFloat,S}
+    num_is_zero = iszero(num)
+    den_is_zero = iszero(den)
+    sb = signbit(num)
+    is_zero = num_is_zero & !den_is_zero
+    is_inf = !num_is_zero & den_is_zero
```

Comment:
I prefer to avoid `return` when possible. It's a jump/goto statement, breaking nice control-flow properties known as **structured programming**. I *do* use jumps in control flow, but only when necessary.

Reply:
I used to be a Go nut, so I see where you're coming from, though.

Reply:
It's only a matter of style of course. I'm not going to fight you over it :smile: